### PR TITLE
Homework make_unique

### PR DESCRIPTION
**Pull request by J-Cieplinski**

### CODE REVIEW

File: module1/homework/make_unique/make_unique.hpp [hunk 1]
```diff
@@ -0,0 +1,14 @@
+#include <memory>
+
+namespace  cs {
+
+    template<typename T, typename... Args, typename = std::enable_if_t<!std::is_array_v<T>>>
```
Comment (majkel84) [inline module1/homework/make_unique/make_unique.hpp:5]: good idea to pass enable_if as template argument

**Review by majkel84 — APPROVED** [review]

**Review by MaciejSalamonski — APPROVED** [review]
i dont have any comments. Good job 👍

File: .github/workflows/make_unique.yaml [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+      - name: mkdir module1/homework/make_unique/build
+        run: mkdir module1/homework/make_unique/build
+      - name: Compile
+        working-directory: module1/homework/make_unique/build
+        run: |
+          cmake ..
+          make
+      - name: Run tests
+        working-directory: module1/homework/make_unique/build
+        run: ctest -V
```
Comment (ziobron) [inline .github/workflows/make_unique.yaml]: you do not check memory leaks with valgrind / asan

File: module1/homework/make_unique/make_unique.hpp [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+    std::unique_ptr <T> make_unique(Args &&... args) {
+        return std::unique_ptr<T>{new T{std::forward<Args>(args)...}};
+    }
+
+    template<typename T>
+    std::unique_ptr <T> make_unique(size_t size) {
+        return std::unique_ptr<T>{new std::remove_extent_t<T>[size] {}};
+    }
```
Comment (ziobron) [inline module1/homework/make_unique/make_unique.hpp]: This should be a proper implementation. Without that memory leaks / undefined behavior is possible.
```suggestion
    template<typename T>
    std::unique_ptr<T[]> make_unique(size_t size) {
        return std::unique_ptr<T[]>{new std::remove_extent_t<T>[size] {}};
    }
```

File: module1/homework/make_unique/test.cpp [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+constexpr int bValue{true};
+
+
+struct TestType {
+    TestType() = default;
+    TestType(int&, double&&, bool) : i_(iValueL), d_(dValueR), b_(bValue) { }
+    TestType(int&&, double&, bool) : i_(iValueR), d_(dValueL), b_(bValue) { }
+
+    int i_ = testI;
+    int d_ = testD;
```
Comment (ziobron) [inline module1/homework/make_unique/test.cpp]: implicit conversion. `testD` is `double`.

File: module1/homework/make_unique/test.cpp [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+    ASSERT_EQ(ptr->d_, dValueR);
+    ASSERT_EQ(ptr->b_, bValue);
+
+    double valueDouble{10.0};
+    auto ptr1 = cs::make_unique<TestType>(0.0, valueDouble, false);
+
+    ASSERT_EQ(ptr1->i_, iValueR);
+    ASSERT_EQ(ptr1->d_, dValueL);
+    ASSERT_EQ(ptr1->b_, bValue);
+}
```
Comment (ziobron) [inline module1/homework/make_unique/test.cpp]: Nice way of checking the behavior.
I rather thought about mock and checking if a proper method was called, but your method is very good 👍

**Review by ziobron — CHANGES_REQUESTED** [review]
Nice tests, but there is no valgrind / asan

File: .github/workflows/make_unique.yaml (hunk 2 above)
Comment (J-Cieplinski) [reply .github/workflows/make_unique.yaml]: I saw no reason to add it. I'm using STL implementation of make_unique

File: module1/homework/make_unique/make_unique.hpp (hunk 3 above)
Comment (J-Cieplinski) [reply module1/homework/make_unique/make_unique.hpp]: Oh, ok

File: module1/homework/make_unique/test.cpp (hunk 4 above)
Comment (J-Cieplinski) [reply module1/homework/make_unique/test.cpp]: I messed up here, `d_` should be `double`

File: module1/homework/make_unique/make_unique.hpp (hunk 3 above)
Comment (J-Cieplinski) [reply module1/homework/make_unique/make_unique.hpp]: However, this is not compiling :D And local valgrind without it showed no leaks

**Review by ziobron — APPROVED** [review]
Good. It looks that I incorrectly suggested replacing T with T[] in array version, because [] is already deduced by compiler and substituted with brackets.

**Comment by ziobron** [comment]
{ "requirement_ids": [1119, 4443], "points": [3, 40] }

File: .github/workflows/make_unique.yaml (hunk 2 above)
Comment (ziobron) [reply .github/workflows/make_unique.yaml]: There is a reason. You used std::unique_ptr but it could be used incorrectly and cause memory leaks.

**Pull request closed without merging**